active_discussion/validity.md

```diff
@@ -0,0 +1,120 @@
+values.
+
+### Extent of "always"
+
+One point we will have to figure out is what exactly "always" means.  Thinking
+in terms of a semantics for MIR, data most probably needs to be valid any time
+it is copied, which primarily happens when executing assignment statements (the
+other cases are passing of function arguments and return values).  However, it
+is less clear whether merely creating a place without accessing the data inside
+(such as in `&*x`) should require the data to be valid.
+
+The entire discussion here is only about validity invariants that have to hold
+when the compiler considers a variable initialized.  For example, `let b: bool;`
+is completely okay to not be initialized because the compiler knows about that;
+`let b: bool = mem::uninitialized();` however copies uninitialized data at type
+`bool` and hence violates `bool`'s validity invariant.
+
+## Goals
+
+* For every primitive type, determine which assumptions (if any) the compiler
+  makes about values *not* occurring at that type (serving as a lower bound for
+  what to declare invalid), and determine which popular patterns in unsafe code
+  might create "interesting" values of this type that safe code cannot create on
+  its own (serving as an upper bound for how much we want to declare invalid).
+  Both of these bounds are soft, but informative.
+* Based on that, map out a design space of invariants that seem reasonable.
+* Determine when exactly the validity invariant is assumed to hold.
+
+## Active threads
+
+To start, we will create threads for each major category of types.
+
+* Integers and floating point types
+  * Do we allow values that contain uninitialized bits?  If yes, what are the
+    rules for arithmetic and logical operations involving uninitialized bits,
+    e.g. in cases like `x * 0`?  There is also some interaction with bug finding
+    here: tools can only flag uninitialized data at integer type as a bug if we
+    do not allow that to happen in unsafe code.
+
+* Raw pointers
+  * Do we allow values that contain uninitialized bits?
+  * Are there any requirements on the metadata?
+
+* References
+  * Presumably, references must be non-NULL.
+  * They probably also must be aligned, but is that required every time a
+    reference is taken?  Also see the [ongoing discussion in RFC 2582][RFC2582].
+  * Can there ever be uninitialized bits in a reference?
+  * Do they have to be dereferencable?  What exactly does that even mean?
+  * Does `&[mut] T` have to point to data that is valid at `T`?  This interacts
+    with the question of whether `&*x` is allowed when `x` is a well-aligned
+    non-null dereferencable pointer that points to invalid data.
+  * Out of scope: aliasing rules
+
+* Function pointers
+  * Presumably, these must be non-NULL.  Anything else?  Can there ever be
+    uninitialized bits?
+
+* Booleans
+  * Is there anything to say besides: A `bool` must be `0x0` or `0x1`?  Do we
+    allow the remaining bits to be uninitialized?
```

Comment:
According to C, ` _Bool` has two values and `sizeof(_Bool) - 2` "trap" representations. We might need to reconcile Rust's "uninitialized" with C's "trap" representation terminology somewhere as part of the validity discussions. IIRC the behavior of C's "trap" representations is undefined, so we can probably define it however we want here, but then that's a requirement on the C platform to play ball.

Reply:
In my mental model, "trap representation" is pretty much equivalent with "indeterminate data" is equivalent with the "uninitialized bits".  I will add a remark along those lines.

Other than that, hold these thoughts until the actual discussion starts. :)

Reply:
What I tried to say is that maybe we should add a question: "Are Rust `uninitialized` bits equivalent to C's `trap` representations ? (example: `_Bool` ...)". I agree with you that the answer is probably yes, but that's something worth triple checking.

Reply:
Given the lack of a formal model of C's trap representations, I think that's pretty much our choice. ;)

Rust `uninitialized` bits are like `poison` in LLVM, and can be soundly used to model `undef` in LLVM (but are not equivalent to `undef`).  That much I can say because `undef` and `poison` are much better specified than trap representations.